File: Components/DevExpressVCL/Demos/ExpressEditors Library/CBuilder/ImageViewerDemo/ImageViewerDemoResizeImage.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "ImageViewerDemoResizeImage.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxButtons"
#pragma link "cxCheckBox"
#pragma link "cxContainer"
#pragma link "cxControls"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxGroupBox"
#pragma link "cxLabel"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxMaskEdit"
#pragma link "cxSpinEdit"
#pragma link "cxTextEdit"
#pragma link "dxCheckGroupBox"
#pragma resource "*.dfm"
TImageViewerDemoResizeImageForm *ImageViewerDemoResizeImageForm;

//---------------------------------------------------------------------------
__fastcall TImageViewerDemoResizeImageForm::TImageViewerDemoResizeImageForm(TComponent* Owner)
	: TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TImageViewerDemoResizeImageForm::cgbPixelsPropertiesChange(TObject *Sender)
{
  seScale->Enabled = ! cgbPixels->CheckBox->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TImageViewerDemoResizeImageForm::teHorizontalPropertiesValidate(TObject *Sender,
  Variant &DisplayValue, TCaption &ErrorText, Boolean &Error)
{
  int AValue = StrToIntDef(VarToStr(DisplayValue), 0);
  Error = (AValue < 1)||(AValue > 10000);
}
//---------------------------------------------------------------------------

void __fastcall TImageViewerDemoResizeImageForm::teWidthPropertiesChange(TObject *Sender)
{
  CalculateHeight();
}
//---------------------------------------------------------------------------

void __fastcall TImageViewerDemoResizeImageForm::teHeightPropertiesChange(TObject *Sender)
{
  CalculateWidth();
}
//---------------------------------------------------------------------------

void __fastcall TImageViewerDemoResizeImageForm::cbAspectRatioPropertiesChange(TObject *Sender)
{
  CalculateHeight();
}
//---------------------------------------------------------------------------

void TImageViewerDemoResizeImageForm::SetGlyphHeight(Integer Value)
{
  if (FGlyphHeight != Value) {
	FGlyphHeight = Value;
	teHeight->Text = IntToStr(Value);
  }
}
//---------------------------------------------------------------------------

void TImageViewerDemoResizeImageForm::SetGlyphWidth(Integer Value)
{
  if (FGlyphWidth != Value) {
	FGlyphWidth = Value;
	teWidth->Text = IntToStr(Value);
  }
}
//---------------------------------------------------------------------------

void TImageViewerDemoResizeImageForm::CalculateHeight()
{
  if (FLockCount == 0 && teWidth->Text != "" && cbAspectRatio->Checked) {
	FLockCount++;
	try {
	  teHeight->Text = IntToStr(div(StrToInt(teWidth->Text) * GlyphHeight, GlyphWidth).quot);
	}
	__finally {
	  FLockCount--;
	}
  }
}
//---------------------------------------------------------------------------

void TImageViewerDemoResizeImageForm::CalculateWidth()
{
  if (FLockCount == 0 && teHeight->Text != "" && cbAspectRatio->Checked) {
	FLockCount++;
	try {
	  teWidth->Text = IntToStr(div(StrToInt(teHeight->Text) * GlyphWidth, GlyphHeight).quot);
	}
	__finally {
	  FLockCount--;
	}
  }
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressEditors Library/Delphi/ImageViewerDemo/ImageViewerDemoResizeImage.pas
unit ImageViewerDemoResizeImage;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, cxGraphics, cxControls, cxLookAndFeels, cxLookAndFeelPainters,
  cxContainer, cxEdit, Menus, StdCtrls, cxButtons, cxTextEdit, cxGroupBox, 
  dxCheckGroupBox, cxMaskEdit, cxSpinEdit, cxLabel, cxCheckBox;

type
  TImageViewerDemoResizeImageForm = class(TForm)
    cxLabel1: TcxLabel;
    seScale: TcxSpinEdit;
    cgbPixels: TdxCheckGroupBox;
    cxLabel2: TcxLabel;
    cxLabel3: TcxLabel;
    teWidth: TcxTextEdit;
    teHeight: TcxTextEdit;
    btnApply: TcxButton;
    cxButton1: TcxButton;
    cbAspectRatio: TcxCheckBox;
    procedure cgbPixelsPropertiesChange(Sender: TObject);
    procedure teHorizontalPropertiesValidate(Sender: TObject;
      var DisplayValue: Variant; var ErrorText: TCaption;
      var Error: Boolean);
    procedure teWidthPropertiesChange(Sender: TObject);
    procedure teHeightPropertiesChange(Sender: TObject);
    procedure cbAspectRatioPropertiesChange(Sender: TObject);
  private
    FGlyphHeight: Integer;
    FGlyphWidth: Integer;
    FLockCount: Integer;

    procedure SetGlyphHeight(Value: Integer);
    procedure SetGlyphWidth(Value: Integer);

    procedure CalculateHeight;
    procedure CalculateWidth;
  public
    property GlyphHeight: Integer read FGlyphHeight write SetGlyphHeight;
    property GlyphWidth: Integer read FGlyphWidth write SetGlyphWidth;
  end;

var
  ImageViewerDemoResizeImageForm: TImageViewerDemoResizeImageForm;

implementation

{$R *.dfm}

procedure TImageViewerDemoResizeImageForm.cgbPixelsPropertiesChange(Sender: TObject);
begin
  seScale.Enabled := not cgbPixels.CheckBox.Checked;
end;

procedure TImageViewerDemoResizeImageForm.teHorizontalPropertiesValidate(Sender: TObject;
  var DisplayValue: Variant; var ErrorText: TCaption; var Error: Boolean);
begin
  try
    StrToInt(DisplayValue);
    Error := StrToInt(DisplayValue) < 1;
  except
    Error := True;
  end;
end;

procedure TImageViewerDemoResizeImageForm.teWidthPropertiesChange(
  Sender: TObject);
begin
  CalculateHeight;
end;

procedure TImageViewerDemoResizeImageForm.teHeightPropertiesChange(
  Sender: TObject);
begin
  CalculateWidth;
end;

procedure TImageViewerDemoResizeImageForm.SetGlyphHeight(Value: Integer);
begin
  if FGlyphHeight <> Value then
  begin
    FGlyphHeight := Value;
    teHeight.Text := IntToStr(Value);
  end;
end;

procedure TImageViewerDemoResizeImageForm.SetGlyphWidth(Value: Integer);
begin
  if FGlyphWidth <> Value then
  begin
    FGlyphWidth := Value;
    teWidth.Text := IntToStr(Value);
  end;
end;

procedure TImageViewerDemoResizeImageForm.CalculateHeight;
begin
  if (FLockCount = 0) and (teWidth.Text <> '') and cbAspectRatio.Checked then
  begin
    Inc(FLockCount);
    try
      teHeight.Text := IntToStr(Round(StrToInt(teWidth.Text) * GlyphHeight / GlyphWidth));
    finally
      Dec(FLockCount);
    end;
  end;
end;

procedure TImageViewerDemoResizeImageForm.CalculateWidth;
begin
  if (FLockCount = 0) and (teHeight.Text <> '') and cbAspectRatio.Checked then
  begin
    Inc(FLockCount);
    try
      teWidth.Text := IntToStr(Round(StrToInt(teHeight.Text) * GlyphWidth / GlyphHeight));
    finally
      Dec(FLockCount);
    end;
  end;
end;

procedure TImageViewerDemoResizeImageForm.cbAspectRatioPropertiesChange(
  Sender: TObject);
begin
  CalculateHeight;
end;

end.
